[AutoFill Debugging] Part 2: Add new WKTextExtraction* types and properties

whsieh opened the pull request:
#### 7cfd9a324a9522418db499e24276d8cdff91232f
<pre>
[AutoFill Debugging] Part 2: Add new WKTextExtraction* types and properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=296966">https://bugs.webkit.org/show_bug.cgi?id=296966</a>

Reviewed by Abrar Rahman Protyasha.

More work in progress towards AutoFill debugging — add several new types of text extraction items,
as well as additional members (which are currently unused). The new types are:

- `WKTextExtractionTextFormControlItem`
- `WKTextExtractionContentEditableItem`
- `WKTextExtractionLinkItem`
- `WKTextExtractionPopupMenu`
- `WKTextExtractionSelectItem`

...and the new properties on all text extraction items are:

- `eventListeners` (representing a set of event listener types on the corresponding element)
- `ariaAttributes` (representing a subset of ARIA attributes)
- `accessibilityRole` (representing the `role` DOM attribute)
- `nodeIdentifier` (representing a nullable identifier string for the node)

* Source/WebCore/page/text-extraction/TextExtractionTypes.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
(eventListenerTypesAsArray(_:)):
(WKTextExtractionContentEditableItem.isFocused):
(WKTextExtractionTextFormControlItem.isSecure):
(WKTextExtractionTextFormControlItem.isFocused):
(WKTextExtractionTextFormControlItem.label):
(WKTextExtractionTextFormControlItem.placeholder):
(WKTextExtractionTextFormControlItem.isReadonly):
(WKTextExtractionTextFormControlItem.isDisabled):
(WKTextExtractionTextFormControlItem.isChecked):
(WKTextExtractionLinkItem.url):
(WKTextExtractionSelectItem.supportsMultiple):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::eventListenerTypes):
(WebKit::createWKEditable):
(WebKit::createItemWithChildren):
(WebKit::createWKTextItem): Deleted.

Canonical link: <a href="https://commits.webkit.org/298299@main">https://commits.webkit.org/298299@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ffdf2504986db89787ab274d7782b70bba0bcbbd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115022 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/34738 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25219 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121142 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/65671 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e5b58998-7f48-45b5-92d9-65328319a424) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/116911 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35383 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43299 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87405 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3ab27769-0b69-40fe-aeec-13e2dad33d83) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117970 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28199 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103264 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67801 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27371 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21384 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/64794 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97580 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21500 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124332 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41995 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31397 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96205 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42367 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99454 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95990 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24433 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41190 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19026 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/38006 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/41870 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47405 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41414 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/44731 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43150 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
```diff
@@ -33,15 +33,76 @@ internal import WebKit_Internal
 // FIXME: Adopt `@objc @implementation` when support for macOS Sonoma is no longer needed.
 // FIXME: (rdar://110719676) Remove all `@objc deinit`s when support for macOS Sonoma is no longer needed.
 
+private func eventListenerTypesAsArray(eventListeners: WKTextExtractionEventListenerTypes) -> [String]? {
```
I think this would be better if an extension computed property was defined on WKTextExtractionEventListenerTypes that conforms it to [CustomStringConvertible](https://developer.apple.com/documentation/Swift/CustomStringConvertible) as well as an `all` computed property and then this function can just become `WKTextExtractionEventListenerTypes.all.compactMap { eventListeners.contains($0) ? String(describing: $0) : nil }`

Also, this function should return a non-optional array

whsieh replied:
> extension computed property was defined on WKTextExtractionEventListenerTypes…

This seems to break open source builds. I'll keep it the way it was.

> Also, this function should return a non-optional array

Done.

rr-codes replied:
ah, then adding the description property with the `fileprivate` access modifier and removing the conformance to the protocol, and then using `$0.description` directly should workaround it. I strongly recommend this, it is much more Swift-y and also improves readability and removes some duplication. (If you do decide to do this, you also don't need any of the explicit `return` keywords)

whsieh replied:
Thanks, giving this a try.

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
```diff
@@ -33,15 +33,76 @@ internal import WebKit_Internal
 // FIXME: Adopt `@objc @implementation` when support for macOS Sonoma is no longer needed.
 // FIXME: (rdar://110719676) Remove all `@objc deinit`s when support for macOS Sonoma is no longer needed.
 
+private func eventListenerTypesAsArray(eventListeners: WKTextExtractionEventListenerTypes) -> [String]? {
```
```suggestion
private static func eventListenerTypesAsArray(_ eventListeners: WKTextExtractionEventListenerTypes) -> [String]? {
```

whsieh replied:
This suggestion also breaks the build.

rr-codes replied:
what's the error message out of curiosity? And did you make the obj-c declaration of this method `+` instead of `-`? (sorry I didn't specify this earlier)

whsieh replied:
```
…/OpenSource/Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:73:16: error: static methods may only be declared on a type
private static func eventListenerTypesAsArray(_ eventListeners: WKTextExtractionEventListenerTypes) -> [String] {
```

> And did you make the obj-c declaration of this method

I did not, because there is no ObjC declaration of this method.

rr-codes replied:
Ohhh sorry didn’t realize this was outside the type. My bad!

rr-codes commented on Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
```diff
@@ -54,9 +115,154 @@ extension WKTextExtractionItem {
+    ) {
         self.container = container
-        super.init(with: rectInWebView, children: children)
+        super
```
does the linter complain if the `.init(` itself is the same line as the super?

whsieh replied:
Yes

whsieh committed:
[AutoFill Debugging] Part 2: Add new WKTextExtraction* types and properties
https://bugs.webkit.org/show_bug.cgi?id=296966

Reviewed by Abrar Rahman Protyasha.

More work in progress towards AutoFill debugging — add several new types of text extraction items,
as well as additional members (which are currently unused). The new types are:

- `WKTextExtractionTextFormControlItem`
- `WKTextExtractionContentEditableItem`
- `WKTextExtractionLinkItem`
- `WKTextExtractionPopupMenu`
- `WKTextExtractionSelectItem`

...and the new properties on all text extraction items are:

- `eventListeners` (representing a set of event listener types on the corresponding element)
- `ariaAttributes` (representing a subset of ARIA attributes)
- `accessibilityRole` (representing the `role` DOM attribute)
- `nodeIdentifier` (representing a nullable identifier string for the node)

* Source/WebCore/page/text-extraction/TextExtractionTypes.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _requestTextExtraction:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtraction.swift:
(eventListenerTypesAsArray(_:)):
(WKTextExtractionContentEditableItem.isFocused):
(WKTextExtractionTextFormControlItem.isSecure):
(WKTextExtractionTextFormControlItem.isFocused):
(WKTextExtractionTextFormControlItem.label):
(WKTextExtractionTextFormControlItem.placeholder):
(WKTextExtractionTextFormControlItem.isReadonly):
(WKTextExtractionTextFormControlItem.isDisabled):
(WKTextExtractionTextFormControlItem.isChecked):
(WKTextExtractionLinkItem.url):
(WKTextExtractionSelectItem.supportsMultiple):
* Source/WebKit/UIProcess/API/Cocoa/_WKTextExtractionInternal.h:
* Source/WebKit/UIProcess/Cocoa/TextExtraction/WKTextExtractionUtilities.mm:
(WebKit::eventListenerTypes):
(WebKit::createWKEditable):
(WebKit::createItemWithChildren):
(WebKit::createWKTextItem): Deleted.

Canonical link: https://commits.webkit.org/298299@main
